altos/ao-tools/ao-dump-up/ao-dump-up.1
.\"
.\" Copyright © 2009 Keith Packard <keithp@keithp.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH AO-DUMPLOG 1 "ao-dump-up" ""
.SH NAME
ao-dump-up \- Dump flight log from MicroPeak flight computer
.SH SYNOPSIS
.B "ao-dump-up"
[\-T \fItty-device\fP]
[\--tty \fItty-device\fP]
[\-D \fIaltos-device\fP]
[\--device \fIaltos-device\fP]
[\--wait]
.SH OPTIONS
.TP
\-T tty-device | --tty tty-device
This selects which tty device ao-dump-up uses to communicate with
the target device.
.TP
\-D AltOS-device | --device AltOS-device
Search for a connected device. This forces the program to look
for a specific USB device name.
.TP
\--wait
Wait for a device to appear instead of exiting when no device is found.
.SH DESCRIPTION
.I ao-dump-up
downloads a MicroPeak flight log from a connected MicroPeak USB adapter.
.SH USAGE
.I ao-dump-up
connects to the specified target device and dumps the stored flight
log.
.SH AUTHOR
Keith Packard


altos/ChangeLog
commit dbcb68f684a96d13efeb9993685f61b27b45e7dc
Author: Bdale Garbee <bdale@gag.com>
Date:   Sat Aug 12 00:21:43 2017 -0400

    update docinfo file for 1.8 release

commit dfb9568866ba250a02039cc3429d5ca0603a60d9
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Aug 12 00:10:23 2017 -0400

    Note that telebt 4.0 files need to go to altusmetrumllc dir
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 6b083e90d5b8de215f2f303f18b5c4ac9185622d
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Aug 12 00:06:48 2017 -0400

    Bump version to 1.8
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 243c1e519d41b45765d4edf5f608655d6ab01eff
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Aug 12 00:06:26 2017 -0400

    doc: Add 1.8 release notes
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 7b67dcdd90317845261e51a1adb2404dcb0f2507
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Aug 12 00:05:50 2017 -0400

    altosui, telegps: Add TeleBT v4.0 firmware to app packages
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 5fdce5b4a70ae59f8ac1ca10d9fadbbbd62412f3
Merge: f1bbf1bd 2a3db289
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Aug 11 23:45:25 2017 -0400

    Merge remote-tracking branch 'origin/master'

commit f1bbf1bd18d52a084af714a022b12bfc3c76374c
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Aug 11 23:40:41 2017 -0400

    ao-bringup: Allow dfu_util to "fail" while flashing chaoskey
    
    Sometimes it isn't happy about the exit status, but if the device
    comes up and works OK, we don't care.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit fbe87bd18308cab62eda6d5e956a8618095ed33b
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Aug 11 23:39:13 2017 -0400

    altoslib: Set time in state for KML output correctly
    
    Use new computed state_time value.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c7b34e6691b632199bc7802c2e4dea4de992578e
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:49:26 2017 -0700

    libaltos: Test both bluetooth and USB APIs with cjnitest
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 52bbfa9cd46ca27a4de937e89d48246a4e94c8a6
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:47:40 2017 -0700

    altos/test: Fix AES key format in test code
    
    Use temp variable instead of stepping on the AES name.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 9eaf917707287af57c2dadfc82d6d415b0964329
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:46:24 2017 -0700

    altos/test: Fix printf format for int64 arithmetic test
    
    Use 'long long' and %lld for 64-bit values when printing.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 198a3d4d5f51de5c0f0413299582479cde4e177f
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:37:27 2017 -0700

    altosui: Rename AltosConfig -> AltosConfigFC
    
    Now the class name says that this configures the flight computer.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit c7c2fc2d85414fefda0a7948a6c4e38f65140861
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Jul 23 16:38:09 2017 -0700

    altosui: Adapt to flight stats time value changes
    
    There aren't state time values anymore as those don't work when you
    have multiple motors. Instead, 'boost_time' is when the rocket left
    the pad and 'landed_time' is when it touched down. Use these new
    values in the --summary output.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 2a3db28953f5cd981c519a917cdcd9f48dcb08e9
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:49:21 2017 -0600

    add TeleBT v4.0 to the list of stuff we build

commit 04560a7cd6a64e742954a2d0fbd7563e2c310402
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:46:55 2017 -0600

    ignore more build products in telebt-v4.0 tree

commit c8dbcaf69cd538a31ab6e2b568237ae7c8656a9a
Merge: 0cbfa444 ea3b5815
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:44:03 2017 -0600

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 0cbfa444a9f9159cb509bb47ca5590fc1d709f64
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:43:38 2017 -0600

    clean up telefireone name references

commit 3a1512b605a81ebee23cc6b7980b709a54cb43c9
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Jul 21 17:42:54 2017 -0600

    rename existing turnon_telebt script, create a new one for v4.0

commit ea3b5815b27005b2f4c3034715f656d28ea8534e
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Jul 21 16:16:55 2017 -0700

    libaltos: Add btle test program
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 545bdc1f8046b08e237307d97d0044ddae9095b8
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Jul 21 16:14:31 2017 -0700

    altos/test: Add EasyMini test mode
    
    Use baro-only mode, parse easymini logs.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 2969f0af52dfbdea7593b208e510c7b17227064c
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Jul 21 16:09:55 2017 -0700

    altoslib: fix adjustment of landed time
    
    Push landed time out if we computed silliness.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 30b308ad0189b4aa9dd5cc7ee876b55fee355a56
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Jul 21 16:05:45 2017 -0700

    Set version to 1.7.1 for TeleBT firmware build
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit b923e25a8a49dda40e62070dd2df6ff9130d4b38
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Jun 27 23:39:16 2017 -0700

    altos/telebt-v4.0: Use HW flow control to BT module
    
    Now that the stmf0 HW flow control works...
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit e64438271e481ca9f9eeb0ce2b1f9eca071b7ad4
Author: Keith Packard <keithp@keithp.com>

altos/ao-tools/ao-dump-up/ao-dump-up.c
/*
 * Copyright © 2009 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <stdbool.h>
#include <ctype.h>
#include "cc-usb.h"
#include "cc.h"

#define NUM_BLOCK	512

static const struct option options[] = {
	{ .name = "tty", .has_arg = 1, .val = 'T' },
	{ .name = "device", .has_arg = 1, .val = 'D' },
	{ .name = "wait", .has_arg = 0, .val = 'w' },
	{ 0, 0, 0, 0},
};

static void usage(char *program)
{
	fprintf(stderr, "usage: %s [--tty <tty-name>] [--device <device-name>] [--wait]\n", program);
	exit(1);
}

static int get_nonwhite(struct cc_usb *cc, int timeout)
{
	int	c;

	for (;;) {
		c = cc_usb_getchar_timeout(cc, timeout);
		putchar(c);
		if (!isspace(c))
			return c;
	}
}

static const uint8_t test_data[] = {
	0xfc, 0xfd, 0xfe, 0xff, 0xf8, 0xf9, 0xfa, 0xfb, 0x40, 0x00, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f,
	0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f,
	0x20, 0x21, 0x22, 0x23, 0x24, 0x25, 0x26, 0x27, 0x28, 0x29, 0x2a, 0x2b, 0x2c, 0x2d, 0x2e, 0x2f,
	0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x3a, 0x3b, 0x3c, 0x3d, 0x3e, 0x3f,
	0x40, 0x41, 0x42, 0x43, 0x44, 0x45, 0x46, 0x47, 0x48, 0x49, 0x4a, 0x4b, 0x4c, 0x4d, 0x4e, 0x4f,
	0x50, 0x51, 0x52, 0x53, 0x54, 0x55, 0x56, 0x57, 0x58, 0x59, 0x5a, 0x5b, 0x5c, 0x5d, 0x5e, 0x5f,
	0x60, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6a, 0x6b, 0x6c, 0x6d, 0x6e, 0x6f,
	0x70, 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77, 0x78, 0x79, 0x7a, 0x7b, 0x7c, 0x7d, 0x7e, 0x7f,
	0x80, 0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87, 0x88, 0x89,
	0x17, 0x08,
};

static bool test_failed;
static int test_pos;

static void
check_test(uint8_t b)
{
	if (test_pos >= sizeof (test_data) || test_data[test_pos++] != b)
		test_failed = true;
}

static uint8_t
get_hexc(struct cc_usb *cc)
{
	int	c = get_nonwhite(cc, 1000);

	if ('0' <= c && c <= '9')
		return c - '0';
	if ('a' <= c && c <= 'f')
		return c - 'a' + 10;
	if ('A' <= c && c <= 'F')
		return c - 'A' + 10;
	fprintf(stderr, "Non-hex char '%c'\n", c);
	exit(1);
}

static int file_crc;

static const int POLY = 0x8408;

static int
log_crc(int crc, int b)
{
	int	i;

	for (i = 0; i < 8; i++) {
		if (((crc & 0x0001) ^ (b & 0x0001)) != 0)
			crc = (crc >> 1) ^ POLY;
		else
			crc = crc >> 1;
		b >>= 1;
	}
	return crc & 0xffff;
}

static uint8_t
get_hex(struct cc_usb *cc)
{
	int	a = get_hexc(cc);
	int	b = get_hexc(cc);
	int	h = (a << 4) + b;

	file_crc = log_crc(file_crc, h);
	check_test(h);
	return h;
}

static int get_32(struct cc_usb *cc)
{
	int	v = 0;
	int	i;
	for (i = 0; i < 4; i++) {
		v += get_hex(cc) << (i * 8);
	}
	return v;
}

static int get_16(struct cc_usb *cc)
{
	int	v = 0;
	int	i;
	for (i = 0; i < 2; i++) {
		v += get_hex(cc) << (i * 8);
	}
	return v;
}

static int swap16(int i)
{
	return ((i << 8) & 0xff00) | ((i >> 8) & 0xff);
}

static int find_header(struct cc_usb *cc)
{
	for (;;) {
		if (get_nonwhite(cc, -1) == 'M' && get_nonwhite(cc, 1000) == 'P')
			return 1;
	}
}

int
main (int argc, char **argv)
{
	struct cc_usb	*cc;
	char		*tty = NULL;
	char		*device = NULL;
	int		c;
	int		nsamples;
	int		i;
	int		crc;
	int		current_crc;
	int		wait = 0;

	while ((c = getopt_long(argc, argv, "wT:D:", options, NULL)) != -1) {
		switch (c) {
		case 'w':
			wait = 1;
			break;
		case 'T':
			tty = optarg;
			break;
		case 'D':
			device = optarg;
			break;
		default:
			usage(argv[0]);
			break;
		}
	}
	if (!tty) {
		for (;;) {
			tty = cc_usbdevs_find_by_arg(device, "FT230X Basic UART");
			if (tty) {
				if (wait) {
					printf("tty is %s\n", tty);
					sleep(1);
				}
				break;
			}
			if (!wait)
				break;
			sleep(1);
		}
	}
	if (!tty)
		tty = getenv("ALTOS_TTY");
	if (!tty)
		tty="/dev/ttyUSB0";
	cc = cc_usb_open(tty);
	if (!cc)
		exit(1);
	find_header(cc);
	file_crc = 0xffff;
	get_32(cc);	/* ground pressure */
	get_32(cc);	/* min pressure */
	nsamples = get_16(cc);	/* nsamples */
	for (i = 0; i < nsamples; i++)
		get_16(cc);	/* sample i */
	current_crc = swap16(~file_crc & 0xffff);
	crc = get_16(cc);	/* crc */
	putchar ('\n');
	if (crc == current_crc) {
		if (!test_failed)
			printf("\033[32mValid MicroTest Data\033[39m\n");
		else
			printf("CRC valid\n");
	}
	else
		printf("CRC invalid\n");
	cc_usb_close(cc);
	exit (0);
}
